### PR TITLE
#8 Timeout steps

### LINKED ISSUE #8: Step to set timeout and reset
As as user I want a step to set the wait time before I use an assert-step in order to control the seconds till the timeout of a condition is over. This can be used for a longer waiting time till an element appears or similar.
As an opposite I also want a step to reset the timeout to the default from the configuration.

- Set timeout <\seconds>
- Reset timeout

### PR DESCRIPTION
**Pull request by mario-s**
Closes #8

### CODE REVIEW

File: docs/STEPS.md [hunk 1]
```diff
@@ -1158,6 +1160,30 @@ Support
 |:-----:|:--------------:|:----------:|
 |   ✔   |        ✔       |      ✔     |
 
+## Set timeout \<seconds>
+
+> \* Set timeout "1"
+
+Sets an explicit timeout for steps which use a specific condition to evaluate as true. This allows the user to overwrite the implicit timeout from the configuration.
```
Comment (derdualist) [inline docs/STEPS.md:1167]:
```suggestion
This step overwrites the property `driver_implicit_timeout` for a number of steps, that support it. For now, those steps are:
* [Assert \<by> = \<by_value> exists](#assert-by--by_value-exists)
* [Assert \<by> = \<by_value> does not exist](#assert-by--by_value-does-not-exist)
* [Save placeholder \<placeholder> from attribute \<attribute> of \<by> = \<by_value>]
* [Assert \<by> = \<by_value> attribute \<attribute> exists](#assert-by--by_value-attribute-attribute-exists)
* [Assert \<by> = \<by_value> attribute \<attribute> contains \<value>](#assert-by--by_value-attribute-attribute-contains-value)
* [Assert \<by> = \<by_value> attribute \<attribute> equals \<value>](#assert-by--by_value-attribute-attribute-equals-value)
* [Assert \<by> = \<by_value> attribute \<attribute> does not contain \<value>](#assert-by--by_value-attribute-attribute-does-not-contain-value)

More steps will be refactored soon to support this configurable timeout.
```

File: docs/STEPS.md [hunk 2]
```diff
@@ -1158,6 +1160,30 @@ Support
...
+
+|Desktop|Android (Chrome)|iOS (Safari)|
+|:-----:|:--------------:|:----------:|
+|   ✔   |        ✔       |      ✔     |
+
+## Reset timeout
+
+> \* Reset timeout
+
+Resets the explicit timeout, so that the implicit one from the configuration is used in steps which wait for a specific condition.
```
Comment (derdualist) [inline docs/STEPS.md:1179]:
```suggestion
Resets the explicit timeout set by [Set timeout \<seconds>](#set-timeout-seconds).
```

File: gauge_web_app_steps/app_context.py [hunk 3]
```diff
@@ -31,6 +31,7 @@ def __init__(self, ctx: ExecutionContext = None) -> None:
         self.image_path = ImagePath(config.get_browser().value, config.is_headless())
         self.images = Images(self.report)
         self.diff_formats = config.get_diff_formats()
+        self.explicit_timeout = None
```
Comment (derdualist) [inline gauge_web_app_steps/app_context.py:34]: This should go into the scenario data_store.
Can you check, if the `self.implicit_timeout` can be removed, too?

File: gauge_web_app_steps/web_app_steps.py [hunk 4]
```diff
@@ -689,6 +689,19 @@ def save_placeholder_from_element_attribute(placeholder_name_param: str, attribu
     data_store.scenario[placeholder_name] = attribute_value
 
 
+@step("Set timeout <seconds>")
+def set_timeout(seconds_param: str):
+    seconds = _substitute(seconds_param)
+    assert seconds.isdigit(),\
```
Comment (derdualist) [inline gauge_web_app_steps/web_app_steps.py:695]: floats should also be supported.

**Review by derdualist — COMMENTED** [review]
Nice. I have some suggestions for the docs

File: docs/STEPS.md [hunk 5]
```diff
@@ -1158,6 +1160,39 @@ Support
...
 |   ✔   |        ✔       |      ✔     |
 
+## Set timeout \<seconds>
+
+> \* Set timeout "1"
+
+This step overwrites the property `driver_implicit_timeout` for a number of steps, that support it. For now, those steps are:
+* [Assert \<by> = \<by_value> exists](#assert-by--by_value-exists)
+* [Assert \<by> = \<by_value> does not exist](#assert-by--by_value-does-not-exist)
+* [Save placeholder \<placeholder> from attribute \<attribute> of \<by> = \<by_value>]
```
Comment (derdualist) [inline docs/STEPS.md:1170]: `[Save placeholder \<placeholder> from attribute \<attribute> of \<by> = \<by_value>](#save-placeholder-placeholder-from-attribute-attribute-of-by--by_value)`
I missed the right part of the link. Pls also double check the other links

File: gauge_web_app_steps/app_context.py (hunk 3 above)
Comment (mario-s) [reply gauge_web_app_steps/app_context.py]: done

File: tests/test_web_app_steps.py [hunk 6]
```diff
@@ -31,7 +31,6 @@ def setUp(self):
         self.app_context.report = Mock()
         data_store.spec[app_context_key] = self.app_context
         self.element = Mock()
-        os.environ["driver_implicit_timeout"] = '0'
```
Comment (derdualist) [inline tests/test_web_app_steps.py:34]: I think this timeout existed here because of unnecessarily long test execution time. Better to leave it.

**Review by derdualist — APPROVED** [review]

**Pull request merged**